### ISSUE TITLE
Overlap between the MegaDepth data used for training and the MegaDepth1500 data used for evaluation

### ISSUE THREAD

**Issue by kengomitsui** [verlab/accelerated_features#118]
First of all, thank you for the excellent repository.
I have one question: does XFeat exclude the MegaDepth1500 data during training?

The pairs in assets/megadepth_1500.json that are read at evaluation include items that are not filtered out by the following operation in megadepth.py:

```
self.pair_infos = [pair_info for pair_info in self.pair_infos
                   if pair_info[1] > min_overlap_score and pair_info[1] < max_overlap_score]
```

—in other words, those with min_overlap_score = 0.3 or higher.

For example, the pair listed in assets/megadepth_1500.json

```
"pair_names": [
  "Undistorted_SfM/0015/images/205993922_4ad25ed386_o.jpg",
  "Undistorted_SfM/0015/images/2912685234_7c4c21ee78_o.jpg"
],
```

corresponds to the indices [356, 898] indicated by npz['pair_infos'][3251] in scene_info_0.1_0.7/0015_0.3_0.5.npz.
The overlap value is 0.3182, so it appears this pair would be included in the default training.

**Comment by zh7117** [comment]
yes, you are right. The test pairs are in cluded in the training.
In loftr,  it has following configuration:
"""
cfg.DATASET.TRAIN_LIST_PATH = f"{TRAIN_BASE_PATH}/trainvaltest_list/train_list.txt"
"""
this file records npzs for training, and it does not contain test scene(0015,0022).  

While this code seems load all the the scene for training.

**Comment by kengomitsui** [comment]
Thanks.
This suggests that the evaluation using MegaDepth1500 in the original xfeat might have been overestimated.

**Comment by guipotje** [comment]
Hello folks,

Thanks for your interest in XFeat!

That’s a good catch, and thank you for taking the time to look at the code and report the potential bug. This issue was introduced during a major refactor of the original paper implementation. Initially, we were using LoFTR’s original dataloader, but to make the code as easy to use and as standalone as possible, we simplified much of it. After that, we retrained the model from scratch, which converged to results very similar to those reported in the paper.

To show how this went undetected, I retrained two versions of XFeat from scratch using the default training script in this codebase:

1. One including the validation scenes (0015, 0022) `.npz`  
2. One removing the validation scenes `.npz` entirely from the training dataloader.

It turns out that including or excluding these scenes makes no difference at all on the MegaDepth1500 benchmark, as shown in the plot below. It is actually a bit funny that the version without validation scenes shows a slight advantage toward the end, likely due to random training dynamics. You can also try this yourself by training two versions of XFeat from scratch as described above.

<img width="4162" height="2966" alt="Image" src="https://github.com/user-attachments/assets/f0089af8-4ce4-4dc0-8770-5234c62e1f0b" />

There are several things that contribute to this result:

- The XFeat backbone is very small and simple, with limited capacity to memorize specific scenes, especially given the large size of the MegaDepth dataset with hundreds of scenes.
- As a local feature extractor CNN, XFeat also exhibits reasonable robustness to scene-specific overfitting, in contrast to transformer-based architectures that have global receptive fields, and although more powerful, can more easily memorize global scene structure.
- We further use a data mixture with COCO synthetic warps, which makes it even harder for the network to perform any kind of memorization.

Nevertheless, I will push a small patch to the dataloader that explicitly excludes the validation scenes, in order to keep it consistent with the paper implementation. Thank you again.

**Comment by wangjunmars-bit** [comment]
I ran into the same issue—and after two months of work, I ended up with a massively overfitted model. @guipotje
